src/redisearch_rs/redis_mock/src/allocator.rs

```diff
@@ -58,21 +74,34 @@ pub extern "C" fn alloc_shim(size: usize) -> *mut c_void {
 /// Safety:
 /// 1. The caller must ensure that pointers returned by this function are freed using `free_shim`.
 pub extern "C" fn calloc_shim(count: usize, size: usize) -> *mut c_void {
-    // Check if size is zero
-    if size == 0 || count == 0 {
-        return std::ptr::null_mut();
-    }
+    let req = match count.checked_mul(size) {
+        Some(n) if n > 0 => n,
+        _ => return ptr::null_mut(),
+    };
 
-    // ensure no overflow
-    let size_without_header = count.checked_mul(size).unwrap();
-    // Safety:
-    // The caller will have to guarantee that `total_size` is > 0.
-    let alloc_zeroed = |total_size| unsafe {
-        alloc_zeroed(Layout::from_size_align(total_size, ALIGNMENT).unwrap()) as *mut c_void
+    let alloc_size = match req.checked_add(HEADER_SIZE) {
+        Some(n) => n,
+        None => return ptr::null_mut(),
     };
-    // Safety:
-    // 1. --> We know count > 0, size > 0 --> size_without_header > 0
-    unsafe { generic_shim(size_without_header, alloc_zeroed) }
+
+    // Safety: `alloc_zeroed` is called with a valid non zero `Layout` and returns
+    // a zero initialized block aligned to `ALIGNMENT` or null on OOM.
+    let base = unsafe { alloc_zeroed(layout_for(alloc_size)) };
+    if base.is_null() {
+        return ptr::null_mut();
+    }
+
+    let h0 = base as *mut usize;
+    // Safety: `h0` is within the allocated header region.
+    unsafe { *h0 = alloc_size };
+
+    let h1 = base.wrapping_add(std::mem::size_of::<usize>()) as *mut usize;
+    // Safety: `h1` is within the allocated header region and disjoint from `h0`.
+    // ... we assign req to h1 purely for diagnostics reasons
```

Comment:
```suggestion
    // Since the header is two words-long, and we only need one word
    // for required metadata (i.e. the actual allocated space),
    // we can use the second word to keep track of additional information
    // that can be handy when troubleshooting, such as the size
    // requested by the user.
```

Same for alloc. Let's try to make comments as easy to understand as possible.